src/mito2/src/memtable/time_partition.rs

```diff
@@ -377,6 +499,67 @@ impl TimePartitions {
         inner.parts.clone()
     }
 
+    /// Find existing partitions that match the bulk data's time range and identify
+    /// any new partitions that need to be created
+    fn find_partitions_by_time_range<'a>(
+        &self,
+        existing_parts: &'a [TimePartition],
+        min: Timestamp,
+        max: Timestamp,
+    ) -> Result<(Vec<&'a TimePartition>, Vec<Timestamp>)> {
+        let mut matching = Vec::new();
+
+        let mut present = HashSet::new();
+        // First find any existing partitions that overlap
+        for part in existing_parts {
+            let Some(part_time_range) = part.time_range.as_ref() else {
+                matching.push(part);
+                return Ok((matching, Vec::new()));
+            };
+
+            if !(max < part_time_range.min_timestamp || min >= part_time_range.max_timestamp) {
+                matching.push(part);
+                present.insert(part_time_range.min_timestamp.value());
+            }
+        }
+
+        // safety: self.part_duration can only be present when reach here.
+        let part_duration = self.part_duration.unwrap();
+        let timestamp_type = self
+            .metadata
+            .time_index_column()
+            .column_schema
+            .data_type
+            .as_timestamp()
+            .unwrap(); // SAFETY: timestamp column must be a timestamp.
```

Comment:
Maybe we can add a method to the RegionMetadata to get the timestamp type

Reply:
fixed in 0f60f36c231f060dc57414789e20e03bfadd1069